Add an API to locate unrecovered shards and their state

areek opened the pull request:
This API provides store information for shard copies of indices.
Store information reports on which nodes shard copies exist, the shard
copy version, indicating how recent they are, and any exceptions
encountered while opening the shard index or from earlier engine failure.

By default, only lists store information for shards that have at least one
unallocated copy. When the cluster health status is yellow, this will list
store information for shards that have at least one unassigned replica.
When the cluster health status is red, this will list store information
for shards, which has unassigned primaries.

Endpoints include shard stores information for a specific index, several
indices, or all:

```bash
curl -XGET 'http://localhost:9200/test/_shard_stores'
curl -XGET 'http://localhost:9200/test1,test2/_shard_stores'
curl -XGET 'http://localhost:9200/_shard_stores'
```

The scope of shards to list store information can be changed through
`status` param. Defaults to 'yellow' and 'red'. 'yellow' lists store information of
shards with at least one unassigned replica and 'red' for shards with unassigned
primary shard.
Use 'green' to list store information for shards with all assigned copies.

```bash
curl -XGET 'http://localhost:9200/_shard_stores?status=green'
```

Response:

The shard stores information is grouped by indices and shard ids.

```bash
{
    ...
   "0": { <1>
        "stores": [ <2>
            {
                "sPa3OgxLSYGvQ4oPs-Tajw": { <3>
                    "name": "node_t0",
                    "transport_address": "local[1]",
                    "attributes": {
                        "enable_custom_paths": "true",
                        "mode": "local"
                    }
                },
                "version": 4, <4>
                "allocation" : "primary" | "replica" | "unused", <6>
                "store_exception": ... <5>
            },
            ...
        ]
   },
    ...
}
```
<1> The key is the corresponding shard id for the store information
<2> A list of store information for all copies of the shard
<3> The node information that hosts a copy of the store, the key
    is the unique node id.
<4> The version of the store copy
<5> The status of the store copy, whether it is used as a
    primary, replica or not used at all
<6> Any exception encountered while opening the shard index or
    from earlier engine failure

closes #10952

areek commented:
@s1monw This is still a WIP in terms of documentation and testing, would appreciate a review.

s1monw commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesUnassignedShardsAction.java:
```diff
@@ -0,0 +1,179 @@
+            this.metaData = metaData;
+            this.listener = listener;
+            this.expectedOps = expectedOps;
+            this.opsCount = new AtomicInteger(0);
```
We have a class called `CountDown.java` for this - check it out it might make things simpler her

areek replied:
Thanks for the pointer, switched to using `CountDown`

s1monw commented on core/src/main/java/org/elasticsearch/gateway/TransportNodesListGatewayStartedShards.java:
```diff
@@ -261,29 +272,40 @@ public String getIndexUUID() {
         public void writeTo(StreamOutput out) throws IOException {
             super.writeTo(out);
             out.writeLong(version);
+            out.writeOptionalString(exception);
```
you can actually serialize the entire exception with `StreamOutput#writeThrowable` it might be better to have the stacktrace too?

areek replied:
That makes more sense, now the entire exception is serialized

s1monw commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesUnassignedShardsAction.java:
```diff
@@ -0,0 +1,179 @@
+        Set<String> requestedIndices = new HashSet<>();
+        requestedIndices.addAll(Arrays.asList(request.indices()));
+        List<ShardId> shardIdsToFetch = new ArrayList<>(state.routingNodes().unassigned().size());
+        for (MutableShardRouting shard : state.routingNodes().unassigned()) {
```
since you are executing this on the master I think you should also take `state.routingNodex().ignoredUnassigned()` into account. Those shards are also interesting. I wonder if we should _flag_ them somehow special but we can do that later...

areek replied:
Now we iterate through all `unassigned` and `ignoredUnassigned` shards.

s1monw commented:
wow @areek this looks pretty awesome. I left some comments

areek commented:
Thanks for the review @s1monw! Addressed all the comments. 

Was wondering if there are any tests that I can look at to get the cluster to have a bunch of unassigned nodes (currently just stoping random nodes)? 
I see `CorruptedFileTest#corruptRandomPrimaryFile`, was wondering if there is a easier way to test out the corruption exception for the response?

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesUnassignedShardsAction.java:
```diff
@@ -0,0 +1,180 @@
+        Set<String> requestedIndices = new HashSet<>();
+        requestedIndices.addAll(Arrays.asList(request.indices()));
+        List<ShardId> shardIdsToFetch = new ArrayList<>();
+        for (MutableShardRouting shard : Iterables.concat(routingNodes.unassigned(), routingNodes.ignoredUnassigned())) {
```
you should be able to use state.routingTable().shardsWithState(ShardRoutingState.UNASSIGNED);

areek replied:
changed to using `state.routingTable().shardsWithState(ShardRoutingState.UNASSIGNED)`

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesUnassignedShardsAction.java:
```diff
@@ -0,0 +1,180 @@
+            }
+
+            void finish() {
+                ImmutableOpenMap.Builder<String, Map<Integer, java.util.List<ShardStatus>>> shardsResponseBuilder = ImmutableOpenMap.builder();
```
we can use IntObjectMap<java.util.List<ShardStatus>> hier for brevity

areek replied:
changed, thanks

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesUnassignedShardsAction.java:
```diff
@@ -0,0 +1,180 @@
+
+            @Override
+            protected synchronized void processAsyncFetch(ShardId shardId, TransportNodesListGatewayStartedShards.NodeGatewayStartedShards[] responses, FailedNodeException[] failures) {
+                super.processAsyncFetch(shardId, responses, failures);
```
I'm doubting about the super call, it just triggers all kind of caching we don't need.

areek replied:
the call to super has been removed

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesUnassignedShardsAction.java:
```diff
@@ -0,0 +1,180 @@
+                    indexShards.put(res.shardId.id(), shardStatuses);
+                    shardsResponseBuilder.put(res.shardId.getIndex(), indexShards);
+                }
+                listener.onResponse(new IndicesUnassigedShardsResponse(shardsResponseBuilder.build()));
```
How do we report failures here?

areek replied:
now, `IndicesUnassigedShardsResponse` takes in an additional `ImmutableList<FailedNodeException>` aggregating all the failures

bleskes commented on core/src/main/java/org/elasticsearch/index/store/Store.java:
```diff
@@ -386,13 +386,24 @@ public static MetadataSnapshot readMetadataSnapshot(Path indexLocation, ESLogger
+     * and the index can be successfully opened. If the index can not
+     * be opened, an exception is thrown
+     */
+    public static boolean tryOpenIndex(Path indexLocation) throws IOException {
```
why do we need the return value? I doesn't add much as we throw an exception. I think it's clearer if we use void.

s1monw replied:
++

areek replied:
changed to void

bleskes commented on core/src/main/java/org/elasticsearch/gateway/TransportNodesListGatewayStartedShards.java:
```diff
@@ -131,16 +141,17 @@ protected NodeGatewayStartedShards nodeOperation(NodeRequest request) {
-        if (metaData == null) {
-            return false;
-        }
+    private boolean tryOpenIndex(ShardId shardId, IndexMetaData metaData) throws IOException {
```
same comment as for Store.tryOpenIndex - I think we should just return void or throw an exception (which we catch higher up anyway)

areek replied:
changed

bleskes commented on core/src/main/java/org/elasticsearch/gateway/TransportNodesListGatewayStartedShards.java:
```diff
@@ -110,9 +111,18 @@ protected NodeGatewayStartedShards nodeOperation(NodeRequest request) {
+                if (metaData != null) {
+                    final boolean canOpenIndex;
+                    try {
+                        canOpenIndex = tryOpenIndex(request.getShardId(), metaData);
```
see comments bellow, we can get rid of canOpenIndex

areek replied:
changed

bleskes commented on core/src/test/java/org/elasticsearch/action/admin/indices/segments/IndicesUnassignedShardsRequestTests.java:
```diff
@@ -0,0 +1,115 @@
+    public void testUnassignedShards() throws Exception {
+        ensureGreen();
+        // TODO: there has to be a better way of having unassigned shards
+        internalCluster().stopRandomDataNode();
```
sadly it's tricky. I don't get something, how do we find the shard copy if we don't have this node in the cluster? Don't we need to restart?

s1monw replied:
I think we can get a reference to the `IndexService->IndexShard` use the `Store` to mark the shard as corrupted and call `failShard` on the `IndexShard` that way we get them _unassigned_

bleskes replied:
that's one way - but it wil only gives us a corrupted shard. Another, which we should test, is disable allocation and restart a node.

bleskes commented:
@areek I think this as awesome api. Left some comments here and there. I think we need to beef up the tests to check for the actual content for the shard responses (check it finds stuff and check that it detects corruption etc.) . I'll respond to the naming part on the ticket..

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesUnassignedShardsAction.java:
```diff
@@ -0,0 +1,186 @@
+                    indexShardsBuilder.put(res.shardId.id(), shardStatuses);
+                    shardsResponseBuilder.put(res.shardId.getIndex(), indexShardsBuilder.build());
+                    for (FailedNodeException failure : res.failures) {
+                        failureBuilder.add(failure);
```
I think we loose the information about which shard has failed. Should we wrap it in DefaultShardOperationFailedException ?

bleskes commented on core/src/test/java/org/elasticsearch/action/admin/indices/segments/IndicesUnassignedShardsRequestTests.java:
```diff
@@ -0,0 +1,116 @@
+public class IndicesUnassignedShardsRequestTests extends ElasticsearchIntegrationTest {
+
+    @Override
+    protected int numberOfReplicas() {
```
why do we override this one? I think we can remove it and leave at defaults, no?

bleskes commented on core/src/test/java/org/elasticsearch/action/admin/indices/segments/IndicesUnassignedShardsRequestTests.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    @Override
+    protected int numberOfShards() {
```
same

bleskes commented:
@areek change looks good. Did you see my comment about beefing up the testing?

areek commented:
@bleskes @clintongormley, I have updated the description with the new API, thoughts? It turned out to be a bit different from what we have discussed before, in terms of default behaviour. It would be good to have this reviewed.

clintongormley commented:
Description in the docs looks good to me

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardsStoresRequest.java:
```diff
@@ -0,0 +1,171 @@
+    /**
+     * Shard criteria to get store information on
+     */
+    public enum ShardState {
```
can this implement Writeable?

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardsStoresRequestBuilder.java:
```diff
@@ -0,0 +1,31 @@
+import org.elasticsearch.action.support.master.MasterNodeReadOperationRequestBuilder;
+import org.elasticsearch.client.ElasticsearchClient;
+
+public class IndicesShardsStoresRequestBuilder extends MasterNodeReadOperationRequestBuilder<IndicesShardsStoresRequest, IndicesShardsStoresResponse, IndicesShardsStoresRequestBuilder> {
```
can we add setters to able to set attributes of the underlying request?

areek replied:
done

bleskes replied:
I still miss the indices, indices options etc?

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardsStoresResponse.java:
```diff
@@ -0,0 +1,362 @@
+        }
+
+        /**
+         * version of the shard, the highest
```
this is not really true - the version is used to select the store that will be used as the very first primary. The version is not used in the choice of promoting an existing replica to a primary

areek replied:
changed, Thanks for catching that.

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardsStoresResponse.java:
```diff
@@ -0,0 +1,362 @@
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject(Fields.NODE);
```
can we take the code from org/elasticsearch/cluster/ClusterState.java:400 (serializing DiscoNodes to xcontent) and put in on DiscoveryNode and reuse it here? (note the tricky node id code)

areek replied:
changed

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardsStoresResponse.java:
```diff
@@ -0,0 +1,362 @@
+        }
+    }
+
+    private ImmutableOpenMap<String, ImmutableOpenIntMap<List<ShardStoreStatus>>> shardStatuses = ImmutableOpenMap.of();
```
move this to the empty constructor?

areek replied:
done

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesShardsStoresAction.java:
```diff
@@ -0,0 +1,220 @@
+            this.fetchResponses = new ConcurrentLinkedQueue<>();
+        }
+
+        void fetch(Collection<ShardId> shardIds) {
```
can we make sure that expectedOps is the size of shardIds by moving the shardIds to the constructor and capturing both?

areek replied:
changed

bleskes commented on core/src/main/java/org/elasticsearch/gateway/TransportNodesListGatewayStartedShards.java:
```diff
@@ -269,29 +275,48 @@ public String getIndexUUID() {
     public static class NodeGatewayStartedShards extends BaseNodeResponse {
 
         private long version = -1;
+        private Throwable exception = null;
```
can we rename this to storeException ?

areek replied:
changed

bleskes commented on core/src/main/java/org/elasticsearch/rest/action/admin/indices/shards/RestIndicesShardStoresAction.java:
```diff
@@ -0,0 +1,66 @@
+    @Inject
+    public RestIndicesShardStoresAction(Settings settings, RestController controller, Client client) {
+        super(settings, controller, client);
+        controller.registerHandler(GET, "/_shard_stores/{shards}", this);
```
did you internationally mean to support GET _shard_stores/unallocated?

areek replied:
oops, thanks for catching that

bleskes commented on core/src/test/java/org/elasticsearch/action/admin/indices/segments/IndicesShardsStoreRequestTests.java:
```diff
@@ -0,0 +1,290 @@
+public class IndicesShardsStoreRequestTests extends ElasticsearchIntegrationTest {
+
+    private void indexRandomData(String index) throws ExecutionException, InterruptedException {
+        int numDocs = scaledRandomIntBetween(100, 1000);
```
does this need to be so big? isn't 10 docs enough?

areek replied:
changed

bleskes commented on core/src/test/java/org/elasticsearch/action/admin/indices/segments/IndicesShardsStoreRequestTests.java:
```diff
@@ -0,0 +1,290 @@
+        }
+
+        logger.info("--> full cluster restart");
+        internalCluster().fullRestart();
```
do you need this restart? it can go, no?

areek replied:
changed

bleskes commented on core/src/test/java/org/elasticsearch/action/admin/indices/segments/IndicesShardsStoreRequestTests.java:
```diff
@@ -0,0 +1,290 @@
+    }
+
+    @Test
+    public void testUnassignedShards() throws Exception {
```
can't this be folded into testBasic? It's practically the same.

areek replied:
folded into testBasic

bleskes commented on core/src/test/java/org/elasticsearch/action/admin/indices/segments/IndicesShardsStoreRequestTests.java:
```diff
@@ -0,0 +1,290 @@
+    }
+
+    @Test
+    public void testSerialization() throws Exception {
```
it seems like this is better done as a unit test?

bleskes replied:
ping?

areek replied:
converted to unit test

bleskes commented:
I went through the code and left some comments - looking good. A couple of other things:

1) I miss the rest spec changes
2) I miss a rest test
3) I think using `shards=allocated` and `shards=unallocatred` as a shard filter is confusing as it is not clear about it's about shard copies that are used or about shards that have any copy allocated. Talked this through with @clintongormley and we settled on having a `status` parameter than can take a comma separated list out of `red,yellow,green,all` and use that to decide which shards will be looked for.
4) "primary_allocated": true in the output seems to be one level too high. What we want to know is whether a store is being used by the cluster or not. As such the information should be on the same level as the store_exception . Naming should also be `allocation: "primary"` or `allocation: "replica"` or `allocation: "unused"` .

Makes sense?

areek commented:
@bleskes would be awesome to have a review on this again. Let me know your thoughts on how the status params determine which shards to get store info for, it should be pretty easy to change this behaviour. Otherwise, I think this is close.

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardsStoresRequest.java:
```diff
@@ -0,0 +1,193 @@
+    }
+
+    private String[] indices = Strings.EMPTY_ARRAY;
+    private IndicesOptions indicesOptions = IndicesOptions.strictExpandOpenAndForbidClosed();
```
why forbid closed indices? we don't really care do we?

areek replied:
The IndicesOptions has been changed to `IndicesOptions.strictExpand()`

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardsStoresRequest.java:
```diff
@@ -0,0 +1,193 @@
+
+    private String[] indices = Strings.EMPTY_ARRAY;
+    private IndicesOptions indicesOptions = IndicesOptions.strictExpandOpenAndForbidClosed();
+    private Status[] shardStatuses = new Status[] { Status.YELLOW };
```
can we use an EnumSet<ClusterHealthStatus> ? We can then let the Status enum go ...right?

bleskes replied:
also the default should be "red,yellow"

areek replied:
changed to EnumSet, thanks for the suggestion

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardsStoresResponse.java:
```diff
@@ -0,0 +1,421 @@
+
+        /**
+         * Version of the shard, used to select the store that will be
+         * used as the very first primary.
```
as the very first primary -> as a primary.

areek replied:
changed

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardsStoresResponse.java:
```diff
@@ -0,0 +1,421 @@
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            node.toXContent(builder, params);
+            builder.field(Fields.VERSION, version);
+            builder.field(Fields.ALLOCATED, allocation.name());
```
does this give you lowercase values? we don't want "allocated": "UNUSED"

areek replied:
Added a `Allocation#value()` method for the string representation instead of using `name()`

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardsStoresResponse.java:
```diff
@@ -0,0 +1,421 @@
+    /**
+     * Single node failure while retrieving shard store information
+     */
+    public static class Failure implements ShardOperationFailedException {
```
maybe inherit from DefaultShardOperationFailedException , to save some code?

areek replied:
Thanks, that did save some code!

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardsStoresResponse.java:
```diff
@@ -0,0 +1,421 @@
+     * Returns {@link ShardStoreStatus}s
+     * grouped by their index names and shard ids.
+     */
+    public ImmutableOpenMap<String, ImmutableOpenIntMap<List<ShardStoreStatus>>> getShardStatuses() {
```
Nit picky - can we call this getStoreStatuses (and all the other ShardStatus relate things should be StoreStatus). Shards are generic and the store describes better what this API does..

areek replied:
makes sense, I removed the Shard prefix for most of the classes/method names in `IndicesShardsStoresResponse `

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesShardsStoresAction.java:
```diff
@@ -0,0 +1,255 @@
+                    routingStates.addAll(Arrays.asList(ShardRoutingState.values()));
+                    break;
+                case YELLOW:
+                case RED:
```
I don't think this is correct?  for yellow we want to have the primary assigned and at least one replica unassigned . for red, the primary should not be assigned..

areek replied:
I think it should make sense now, I knew it cannot be correct, it didn't make a lot of sense to do the same thing for `GREEN`, `ALL` and `YELLOW`, `RED`.

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesShardsStoresAction.java:
```diff
@@ -0,0 +1,255 @@
+            switch (status) {
+                case GREEN:
+                case ALL:
+                    routingStates.addAll(Arrays.asList(ShardRoutingState.values()));
```
Green != ALL. Green are all the shard for which all copies are assigned.

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesShardsStoresAction.java:
```diff
@@ -0,0 +1,255 @@
+            void finish() {
+                Map<String, IntObjectHashMap<java.util.List<NodeGatewayStartedShards>>> shardsResponses = new HashMap<>();
+                ImmutableList.Builder<IndicesShardsStoresResponse.Failure> failures = ImmutableList.builder();
+                while (!fetchResponses.isEmpty()) {
```
can we just do foreach (Repsonse res: fetchResponses) ?

areek replied:
changed

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesShardsStoresAction.java:
```diff
@@ -0,0 +1,255 @@
+                        failures.add(new IndicesShardsStoresResponse.Failure(failure.nodeId(), res.shardId.getIndex(), res.shardId.id(), failure.getCause()));
+                    }
+                }
+                ImmutableOpenMap.Builder<String, ImmutableOpenIntMap<java.util.List<IndicesShardsStoresResponse.ShardStoreStatus>>> indicesShardStoreStatuses = ImmutableOpenMap.builder();
```
this confuses me.. why does it need to be a separate loop? can we inline it with the previous one?

areek replied:
changed

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesShardsStoresAction.java:
```diff
@@ -0,0 +1,255 @@
+                        for (int i = 0; i < shardResponse.value.size(); i++) {
+                            NodeGatewayStartedShards res = shardResponse.value.get(i);
+                            final IndicesShardsStoresResponse.ShardStoreStatus.Allocation allocation;
+                            if (res.storeException() == null) {
```
why should we care about the store exception here? I would just expose the RoutingNodes opinion of this shard & node.

areek replied:
Thanks for the suggestion, this cleared up a lot of the confusion for me

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesShardsStoresAction.java:
```diff
@@ -0,0 +1,255 @@
+                            NodeGatewayStartedShards res = shardResponse.value.get(i);
+                            final IndicesShardsStoresResponse.ShardStoreStatus.Allocation allocation;
+                            if (res.storeException() == null) {
+                                allocation = isPrimary(index, shardResponse.key, res.getNode())
```
FYI - not all shard stores are used .. you can't assume that if there was no error and that it's not a primary it must be a replica.

areek replied:
Now we use `RoutingNodes` to figure out whether it is a primary or replica..

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesShardsStoresAction.java:
```diff
@@ -0,0 +1,255 @@
+            }
+
+            private boolean isPrimary(String index, int shardID, DiscoveryNode node) {
+                IndexRoutingTable indexRoutingTable = routingTables.index(index);
```
I would use routing nodes here - it is keyed by node and will make the lookups more efficient (there is only one copy of a shard per node)

areek replied:
changed

bleskes commented on core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java:
```diff
@@ -468,6 +470,16 @@ public Delta emptyDelta() {
 
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject("nodes");
```
this is tricky - DiscoveryNodes carries more info then just the nodes - it also tells you which one is local and which one is master. If we do have a ToXContent implementation this should be represented - but you may not want to expose it everywhere. Since we only use it in one place now, maybe just roll it back and remove this?

areek replied:
reverted implementing `ToXContext` for `DiscoveryNodes`

bleskes commented on core/src/test/java/org/elasticsearch/action/admin/indices/segments/IndicesShardsStoreRequestTests.java:
```diff
@@ -0,0 +1,275 @@
+                        .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, "2")
+                        .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, "1")
+        ));
+        ensureGreen(index);
```
this one is not needed. The indexing will wait for the shards they need.

areek replied:
removed

bleskes commented on core/src/test/java/org/elasticsearch/action/admin/indices/segments/IndicesShardsStoreRequestTests.java:
```diff
@@ -0,0 +1,275 @@
+        logger.info("--> disable allocation");
+        disableAllocation(index);
+        logger.info("--> stop random node");
+        internalCluster().stopRandomNode(new IndexNonMasterNodePredicate(index, internalCluster().getMasterName()));
```
why do we care about the master here?

areek replied:
`TransportIndicesShardsStoresAction` is implemented as a master-read action. I was under the impression that if the master node is stopped, then the `shardsStores` request will fail? Isn't there any chance that we hit `shardsStores` before a new master is elected?

bleskes commented on core/src/test/java/org/elasticsearch/action/admin/indices/segments/IndicesShardsStoreRequestTests.java:
```diff
@@ -0,0 +1,275 @@
+        disableAllocation(index);
+        logger.info("--> stop random node");
+        internalCluster().stopRandomNode(new IndexNonMasterNodePredicate(index, internalCluster().getMasterName()));
+        ensureYellow(index);
```
why is this needed? we just stopped one node, -> we know it's yellow.

areek replied:
removed

bleskes commented on core/src/test/java/org/elasticsearch/action/admin/indices/segments/IndicesShardsStoreRequestTests.java:
```diff
@@ -0,0 +1,275 @@
+        response = client().admin().indices().shardsStores(Requests.indicesShardsStoresRequest(index).shardStatuses("yellow")).get();
+        assertThat(response.getShardStatuses().containsKey(index), equalTo(true));
+        ImmutableOpenIntMap<List<IndicesShardsStoresResponse.ShardStoreStatus>> shardStoresStatuses = response.getShardStatuses().get(index);
+        assertThat(shardStoresStatuses.size(), equalTo(unassigned.size()));
```
It is technically correct, but only because we have 1 replica and we stopped one node. In general unassigned can have multiple copies of the same shard which may and may not have a store somewhere. Can we just check how many shards have unassigned copies via the routingTable?

bleskes replied:
I think we can also check here that the primary store is returned (we should still have two per shard)

areek replied:
> It is technically correct, but only because we have 1 replica and we stopped one node. In general unassigned can have multiple copies of the same shard which may and may not have a store somewhere.

`TransportIndicesShardsStoresAction` operates on unique `ShardID`s rather than unique `ShardRouting`s and then it gets store info for the unique `ShardID`s from all nodes. So the expected shard entries with store information should be the same as unique `ShardID`s?

> I think we can also check here that the primary store is returned (we should still have two per shard)

Thanks for the suggestion, added the checks

bleskes commented on core/src/test/java/org/elasticsearch/action/admin/indices/segments/IndicesShardsStoreRequestTests.java:
```diff
@@ -0,0 +1,275 @@
+    }
+
+    @Test
+    public void testDefault() throws Exception {
```
do we need this test? isn't it covered by testBasic ?

areek replied:
removed

bleskes commented on core/src/test/java/org/elasticsearch/action/admin/indices/segments/IndicesShardsStoreRequestTests.java:
```diff
@@ -0,0 +1,275 @@
+        assertAcked(prepareCreate(index2).setSettings(Settings.builder()
+                        .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, "2")
+        ));
+        ensureGreen();
```
this one can go away

areek replied:
removed

bleskes commented on core/src/test/java/org/elasticsearch/action/admin/indices/segments/IndicesShardsStoreRequestTests.java:
```diff
@@ -0,0 +1,275 @@
+        logger.info("--> disable allocation");
+        disableAllocation(index);
+
+        logger.info("--> corrupt all shards");
```
do we want to randomly corrupt only some of the shards?

areek replied:
now we randomly corrupt some shards instead

bleskes commented on core/src/test/java/org/elasticsearch/action/admin/indices/segments/IndicesShardsStoreRequestTests.java:
```diff
@@ -0,0 +1,275 @@
+        }
+    }
+
+    private void disableAllocation(String index) {
```
maybe upgrade these to the base class? seems useful...

areek replied:
upgraded!

bleskes commented:
Thx @areek . Left some more comments..

areek commented:
@bleskes Thanks for the review, addressed all the comments. another review?

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardsStoresRequest.java:
```diff
@@ -0,0 +1,209 @@
+/**
+ * Request for {@link IndicesShardsStoresAction}
+ */
+public class IndicesShardsStoresRequest extends MasterNodeReadRequest<IndicesShardsStoresRequest> implements IndicesRequest.Replaceable {
```
nit: can we rename this class and it's family to use a single shard in the name (like the rest `_shard_stores` naming) ? So IndicesShardStoresReqeuest (no Shards)

areek replied:
changed `IndicesShardsStores*` to `IndicesShardStores*`

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardsStoresRequest.java:
```diff
@@ -0,0 +1,209 @@
+    /**
+     * Status used to choose shards to get store information on
+     */
+    public enum Status {
```
do we still need this? Can't we use ClusterHealthStatus ? now that we have the EnumSet , we don't need ALL anymore...

areek replied:
Thanks for the suggestion, we now use ClusterHealthStatus

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardsStoresResponse.java:
```diff
@@ -0,0 +1,383 @@
+        }
+
+        /**
+         * Node the shard belongs to
```
shard -> store

areek replied:
changed

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardsStoresResponse.java:
```diff
@@ -0,0 +1,383 @@
+        }
+
+        /**
+         * Version of the shard, used to select the store that will be
```
shard -> store

areek replied:
changed

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardsStoresResponse.java:
```diff
@@ -0,0 +1,383 @@
+
+        @Override
+        public int compareTo(StoreStatus other) {
+            if (other == null) {
```
Out of curiosity is null a valid incoming value here? Is it required by the compareTo semantics and if so, does null need to be the highest value?

areek replied:
null is not a valid incoming value, removed null check

bleskes commented on core/src/main/java/org/elasticsearch/gateway/TransportNodesListGatewayStartedShards.java:
```diff
@@ -139,16 +143,17 @@ protected NodeGatewayStartedShards nodeOperation(NodeRequest request) {
+        try {
+            Store.tryOpenIndex(shardPath.resolveIndex());
+        } catch (Exception exception) {
+            logger.trace("Can't open index for path [{}]", exception, shardPath.resolveIndex());
```
I think this results in double logging, we will log this here too, no? https://github.com/elastic/elasticsearch/pull/11545/files#diff-9a7b643d56430763f42de913657ca798R125

areek replied:
merged to a single log line

bleskes commented on core/src/test/java/org/elasticsearch/action/admin/indices/segments/IndicesShardsStoreRequestTests.java:
```diff
@@ -0,0 +1,222 @@
+        String index = "test";
+        internalCluster().ensureAtLeastNumDataNodes(2);
+        assertAcked(prepareCreate(index).setSettings(Settings.builder()
+                        .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, "5")
```
do we need so many shards?

bleskes commented on core/src/test/java/org/elasticsearch/action/admin/indices/segments/IndicesShardsStoreResponseTest.java:
```diff
@@ -0,0 +1,123 @@
+        List<IndicesShardsStoresResponse.StoreStatus> storeStatusList = new ArrayList<>();
+        storeStatusList.add(new IndicesShardsStoresResponse.StoreStatus(node1, 3, IndicesShardsStoresResponse.StoreStatus.Allocation.PRIMARY, null));
+        storeStatusList.add(new IndicesShardsStoresResponse.StoreStatus(node2, 2, IndicesShardsStoresResponse.StoreStatus.Allocation.REPLICA, null));
+        storeStatusList.add(new IndicesShardsStoresResponse.StoreStatus(node1, -1, IndicesShardsStoresResponse.StoreStatus.Allocation.UNUSED, new IOException("corrupted")));
```
Is this possible? do we even open the store of the version is -1?

areek replied:
This `StoreStatus` represents when we fail to open the shard index due to some corruption.

bleskes commented:
I left some final minor comments. I think we are getting close!

areek commented:
@bleskes Thanks for the review, addressed all your comments

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardsStoreRequestBuilder.java:
```diff
@@ -0,0 +1,62 @@
+/*
```
Left over IndicesShard*S* ..

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesShardStoresAction.java:
```diff
@@ -0,0 +1,248 @@
+
+        // collect relevant shard ids of the requested indices for fetching store infos
+        for (String index : concreteIndices) {
+            for (IndexShardRoutingTable routing : routingTables.index(index)) {
```
Another simplification - if we push the code at https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/action/admin/cluster/health/ClusterIndexHealth.java#L81 into ClusterIndexShardHealth's constructor, we can use it here and just make it a simple lookup in the enum set..

areek replied:
Thanks for the suggestion.
Now, we use `ClusterShardHealth` to figure out the shard health status in `TransportIndicesShardStoresAction`

bleskes commented on core/src/main/java/org/elasticsearch/client/IndicesAdminClient.java:
```diff
@@ -219,6 +222,29 @@
+     * @return The result future
+     * @see Requests#indicesShardsStoresRequest(String...)
+     */
+    ActionFuture<IndicesShardStoresResponse> shardsStores(IndicesShardStoresRequest request);
```
shard*s*Stores -> shardStores

areek replied:
my bad, missed some renaming, changed

bleskes commented on core/src/main/java/org/elasticsearch/client/IndicesAdminClient.java:
```diff
@@ -219,6 +222,29 @@
+     * @param listener A listener to be notified with a result
+     * @see Requests#indicesShardsStoresRequest(String...)
+     */
+    void shardsStores(IndicesShardStoresRequest request, ActionListener<IndicesShardStoresResponse> listener);
```
same

bleskes commented on core/src/main/java/org/elasticsearch/client/support/AbstractClient.java:
```diff
@@ -1495,6 +1499,21 @@ public IndicesSegmentsRequestBuilder prepareSegments(String... indices) {
         }
 
         @Override
+        public ActionFuture<IndicesShardStoresResponse> shardsStores(IndicesShardStoresRequest request) {
```
shard*s*Stores

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardStoresRequest.java:
```diff
@@ -0,0 +1,128 @@
+     * @param shardStatuses acceptable values are "green", "yellow" and "red"
+     */
+    public IndicesShardStoresRequest shardStatuses(String... shardStatuses) {
+        statuses = EnumSet.noneOf(ClusterHealthStatus.class);
```
can we add a shortcut for `all`? we lost it...

areek replied:
added back the shortcut for `all`

bleskes commented on core/src/main/java/org/elasticsearch/client/Requests.java:
```diff
@@ -185,6 +186,15 @@ public static IndicesSegmentsRequest indicesSegmentsRequest(String... indices) {
+     * @return The indices shard stores request
+     * @see org.elasticsearch.client.IndicesAdminClient#shardsStores(IndicesShardStoresRequest)
+     */
+    public static IndicesShardStoresRequest indicesShardsStoresRequest(String... indices) {
```
shard*S*Stores

bleskes commented on core/src/main/java/org/elasticsearch/gateway/TransportNodesListGatewayStartedShards.java:
```diff
@@ -118,9 +119,18 @@ protected NodeGatewayStartedShards nodeOperation(NodeRequest request) {
+                    try {
+                        shardPath = ShardPath.loadShardPath(logger, nodeEnv, shardId, metaData.settings());
+                        if (shardPath == null) {
+                            throw new IllegalStateException(shardId + " no shard state found");
```
Is this message correct? it should be no shard path found , right?

areek replied:
fixed

bleskes commented on core/src/main/java/org/elasticsearch/gateway/TransportNodesListGatewayStartedShards.java:
```diff
@@ -118,9 +119,18 @@ protected NodeGatewayStartedShards nodeOperation(NodeRequest request) {
+                        Store.tryOpenIndex(shardPath.resolveIndex());
+                    } catch (Exception exception) {
+                        logger.trace("{} can't open index for shard [{}] in path [{}]", exception, shardId, shardStateMetaData, (shardPath != null) ? shardPath.resolveIndex() : "");
+                        return new NodeGatewayStartedShards(clusterService.localNode(), -1, exception);
```
can we report the right version we found ? note that we would probably need to change the the logic in the gateway allocator to check for both -1 version and exception (now -1 means both).

areek replied:
now we report the correct version even in case of existing store exception. `GatewayAllocator` treats shards with store exception the same way as it does non-existent shards on node.

bleskes commented:
Left some final suggestions. Thx @areek

areek commented:
@bleskes Thanks for the review, updated the PR addressing all your comments.

bleskes commented:
LGTM. Left some very minor comment. Thx for all the hard word :+1:

areek commented:
merged to master https://github.com/elastic/elasticsearch/commit/7a21d846bb920b25101a2a958dbfb44f872cfeb1
